File: leetcode/lc-us/cpp/1449.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string largestNumber(vector<int> &cost, int target) {
        vector<int> f(target + 1, -1e8);
        f[0] = 0;
        for (int i = 0; i < 9; i++) {
            for (int j = cost[i]; j <= target; j++) {
                f[j] = max(f[j], f[j - cost[i]] + 1);
            }
        }
        if (f[target] <= 0) return "0";
        string res;
        for (int i = 8, j = target; i >= 0; i--) {
            while (j >= cost[i] && f[j] == f[j - cost[i]] + 1) {
                j -= cost[i];
                res += to_string(i + 1);
            }
        }
        return res;
    }
};

File: leetcode/lc-us/py/1449.py
from typing import List


class Solution:
    def largestNumber(self, cost: List[int], target: int) -> str:
        f = [float("-inf")] * (target + 1)
        f[0] = 0
        for c in cost:
            for x in range(c, target + 1):
                f[x] = max(f[x], f[x - c] + 1)
        if f[target] <= 0:
            return "0"
        res = ''
        j = target
        for i in range(8, -1, -1):
            while j >= cost[i] and f[j] == f[j - cost[i]] + 1:
                res += str(i + 1)
                j -= cost[i]
        return res
